### DISCUSSION TITLE
Headless mode - How to commit changes without exclusive checkout

### DISCUSSION THREAD

**Discussion by dzik87** [NationalSecurityAgency/ghidra#4373] [Q&A]
Hi, 

I run automated sync script for one of our project files and have problem when someone on GUI version browse changes for example and forget to undo checkout.

I would like to know if there is any way to force commit even if file is checked out by someone else or at least force undo checkout on this file before checking out by headless.

Thinking about how headless mode works it make not much sense since on GUI version multiple people can work on one file at the same time without problem but in headless mode it is required to have exclusive checkout...

Looking for some advises how to approach this problem.

Thanks in advance.

**Comment by ghidra1** [comment]
The exclusive checkout requirement is done to avoid the need to merge during checkin.  If headless is run by a user with ADMIN repository rights, it is possible to first run a script which could force removal of all checkouts on all project files to be re-processed.  This would need to be run in advance of the re-process invocation.  If you do not have a need to checkin modifications the checkout can be avoided by using the `-readOnly` option.

**Reply by dzik87** [reply]
Thanks for reply. I have posibility to create admin rights account for this repository, so force remove checkouts would be enough to achieve my goal.

I did check docs searching for that but Ii could not find anything regarding this.

Could you maybe advise how to achieve force remove checkouts in headles mode?

Thanks in advance.

**Comment by ghidra1** [comment]
The following code snippet can help you with the associated API calls.  This example does the entire shared project, although it could easily be limited to a specific project folder.  (NOTE: I have not tested this but should work):
```
	private int terminateCheckouts(DomainFolder folder, RepositoryAdapter repo, TaskMonitor monitor) throws IOException, CancelledException {
		int count = 0;
		for (DomainFile df : folder.getFiles()) {
			monitor.checkCanceled();
			if (!df.isVersioned()) {
				continue;
			}
			for (ItemCheckoutStatus checkout : df.getCheckouts()) {
				// println(...log details of termination...);
				repo.terminateCheckout(df.getParent().getPathname(), df.getName(), checkout.getCheckoutId(), true);
				++count;
			}
		}
		
		for (DomainFolder subfolder : folder.getFolders()) {
			monitor.checkCanceled();
			count += terminateCheckouts(subfolder, repo, monitor);
		}
		return count;
	}

	@Override
	protected void run() throws Exception {

		ProjectData projectData = state.getProject().getProjectData();
		
		RepositoryAdapter repo = projectData.getRepository();
		if (repo == null) {
			return; // not a shared project
		}
		
		int count = terminateCheckouts(projectData.getRootFolder(), repo, monitor);
		
		println("Terminated " + count + " checkouts");
	}	

```

**Reply by dzik87** [reply]
Oh i was thinking it would be in clientless command but it can be achieved with python prescript. I'll try to write some code at weekend and share results here.

Thanks
